### PR TITLE
--Add Configuration State Flags - Part 2 (Hidden file path fields)

### PR DESCRIPTION
**Pull request by jturner65**
## Motivation and Context
This PR is the 2nd of 2 PRs started with [this PR](https://github.com/facebookresearch/habitat-sim/pull/2396) that add state-based support to ConfigValues.  This PR is primarily concerned with implementing hidden/internal configuration fields intended to be used for data that is only locally relevant, such as fully-qualified file paths. 

We do not want file names saved to json that embed absolute path data, so we need to keep a version of the path data that remains relative to the location of the JSON config file while we also want to have an easily consumed absolute path to the assets in question.  Using the hidden field feature, paths can be saved in the configs as they are loaded, or otherwise relative to the dataset config directory, while they can be accessed internally fully qualified.

## How Has This Been Tested
Local c++ and python tests pass

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/metadata/attributes/SemanticAttributes.h [hunk 1]
```diff
@@ -179,37 +179,78 @@ class SemanticAttributes : public AbstractAttributes {
...
+   * the hierharchy of semantic information embedded in the Semantic Asset mesh.
+   * May be overridden by value specified in Scene Instance Attributes. This is
+   * only used internally and should not be saved to disk.
+   */
+  std::string getSemanticDescriptorFullPath() const {
+    return get<std::string>("__semanticDescriptorFullPath");
+  }
+
+  /**
+   * @brief Set the retlative Filename to the semantic texture mesh, if one
```
Comment (aclegg3) [inline src/esp/metadata/attributes/SemanticAttributes.h:221]:
```suggestion
   * @brief Set the relative Filename to the semantic texture mesh, if one
```

File: src/esp/metadata/managers/AOAttributesManager.cpp [hunk 2]
```diff
@@ -182,37 +185,49 @@ AOAttributesManager::preRegisterObjectFinalize(
...
   if (urdfFilePath.empty()) {
     // URDF Filepath empty is bad
     ESP_ERROR(Mn::Debug::Flag::NoSpace)
         << "ArticulatedObjectAttributes template named `" << AOAttributesHandle
         << "` does not specify a valid URDF Filepath, so registration is "
            "aborted.";
     return core::managedContainers::ManagedObjectPreregistration::Failed;
-  } else if (!Cr::Utility::Path::exists(urdfFilePath)) {
+  } else if (!Cr::Utility::Path::exists(urdfFilePath) &&
+             !Cr::Utility::Path::exists(urdfFullFilePath)) {
```
Comment (aclegg3) [inline src/esp/metadata/managers/AOAttributesManager.cpp:197]: Do we need to check both? Shouldn't the full path be enough after construction?

File: src/esp/metadata/managers/AttributesManagerBase.h [hunk 3]
```diff
@@ -235,7 +236,54 @@ class AttributesManager : public ManagedFileBasedContainer<T, Access> {
...
+   * the paths specified in an attributes are properly configured.
+   * @param attributes The attributes to be filtered.
+   */
+  virtual void finalizeAttrPathsBeforeRegister(
+      const AttribsPtr& attributes) const = 0;
+
  protected:
+  /**
+   * @brief Called internally right before attribute registration. Filepaths
+   * in the json configs for Habitat Datasets are specified relative to the
```
Comment (aclegg3) [inline src/esp/metadata/managers/AttributesManagerBase.h:252]:
```suggestion
   * in the json configs for Habitat SceneDatasets are specified relative to the
```

File: src/esp/metadata/managers/AttributesManagerBase.h [hunk 4]
```diff
@@ -235,7 +236,54 @@ class AttributesManager : public ManagedFileBasedContainer<T, Access> {
...
+   * This function will examine the @p curRelPathName to see that it is a
+   * relative path in relation to the @p attributes' set filepath, and if not it
+   * will strip out the path information and resave the name to @p rePathSetter.
+   * It will query the passed @p curFQPathName to see that it is a fully
+   * qualified path to a file that exists, and if not it will add the
+   * attributes' filepath and resave it to @p fqPathSetter.
+   *
+   *
+   * It is assumed that what is passed to this function always referneces a
+   * filepath, so if, for example, this is called on a render asseet filepath,
```
Comment (aclegg3) [inline src/esp/metadata/managers/AttributesManagerBase.h:267]:
```suggestion
   * filepath, so if, for example, this is called on a render asset filepath,
```

**Review by aclegg3 — APPROVED** [review]
Looks good to me. Minor typos and a question.

File: src/esp/metadata/managers/AOAttributesManager.cpp (hunk 2 above)
Comment (jturner65) [reply src/esp/metadata/managers/AOAttributesManager.cpp:197]: The full path might not exist yet, depending on what pathway the attributes took to get here.

**Pull request merged**